server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/aws/AwsS3ObjectStorage.java

```diff
@@ -146,32 +146,28 @@ private ContextBuilder contextBuilder() {
         }
 
         @Override
-        public void putDirectly(ObjectStorageBucketName bucketName, Blob blob) {
-            writeFileAndAct(blob, (file) -> putWithRetry(bucketName, configuration, blob, file, FIRST_TRY).block());
+        public Mono<Void> putDirectly(ObjectStorageBucketName bucketName, Blob blob) {
+            return writeFileAndAct(blob, file -> putWithRetry(bucketName, configuration, blob, file, FIRST_TRY));
         }
 
         @Override
-        public BlobId putAndComputeId(ObjectStorageBucketName bucketName, Blob initialBlob, Supplier<BlobId> blobIdSupplier) {
-            Consumer<File> putChangedBlob = (file) -> {
+        public Mono<BlobId> putAndComputeId(ObjectStorageBucketName bucketName, Blob initialBlob, Supplier<BlobId> blobIdSupplier) {
+            Function<File, Mono<Void>> putChangedBlob = file -> {
                 initialBlob.getMetadata().setName(blobIdSupplier.get().asString());
-                putWithRetry(bucketName, configuration, initialBlob, file, FIRST_TRY).block();
+                return putWithRetry(bucketName, configuration, initialBlob, file, FIRST_TRY);
             };
-            writeFileAndAct(initialBlob, putChangedBlob);
-            return blobIdSupplier.get();
+            return writeFileAndAct(initialBlob, putChangedBlob)
+                .then(Mono.defer(() -> Mono.just(blobIdSupplier.get())));
         }
 
-        private void writeFileAndAct(Blob blob, Consumer<File> putFile) {
-            File file = null;
+        private Mono<Void> writeFileAndAct(Blob blob, Function<File, Mono<Void>> putFile) {
             try {
-                file = File.createTempFile(UUID.randomUUID().toString(), ".tmp");
+                File file = File.createTempFile(UUID.randomUUID().toString(), ".tmp");
                 FileUtils.copyToFile(blob.getPayload().openStream(), file);
```

Comment:
You have to wrap all non-trivial operations to Mono, otherwise, you are lying to your caller